libutils/string_lib.c

```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
 
 /*********************************************************************/
 
-long StringToLong(const char *str)
+/**
+ * @brief Converts a string of numerals in base 10 to a long integer.
+ *
+ * Result is stored in *value_out, return value should be checked.
+ * On error *value_out is unmodified and an error code is returned.
+ * Leading spaces in input string are skipped.
+ * String numeral must be terminated by NULL byte or space (isspace).
+ *
+ * @see StringToLongExitOnError()
+ * @see StringToLongDefaultOnError()
+ * @param[in] str String with numerals to convert, cannot be NULL
+ * @param[out] value_out Where to store result on success, cannot be NULL
+ * @return 0 on success, error code otherwise (see source code)
+ */
+int StringToLong(const char *str, long *value_out)
+{
+    assert(str != NULL);
+    assert(value_out != NULL);
+    assert(ERANGE != 0);
+
+    char *endptr = NULL;
+    long val;
+
+    errno = 0;
+    val = strtol(str, &endptr, 10);
+
+    if ((errno == ERANGE && (val == LONG_MAX || val == LONG_MIN)))
+    {
+        return ERANGE; // Overflow or underflow
+    }
+
+    if (endptr == str)
+    {
+        return -81; // No digits found
+    }
+
+    if (endptr == NULL)
+    {
+        return -82; // endpointer not set by strtol
+    }
+
+    if (*endptr != '\0' && !isspace(*endptr))
+    {
+        return -83; // string not properly terminated
+    }
+
+    if (errno != 0)
+    {
+        return errno; // Unknown error
+    }
+
+    *value_out = val;
+    return 0;
+}
+
+/**
+ * @brief Log a conversion error with an identifier for debugging
+ */
+void LogConversionError(const char *str_attempted, const char *id)
+{
+    Log(LOG_LEVEL_ERR, "Conversion error on '%s' (%s)", str_attempted, id);
```

Comment:
Would be nice if this also took the return code and printed the respective error message based on it.

Reply:
That would make it specific to StringToLong, but I agree. I'll add this.